# October 2024

## 03

### Meaningful try-catch blocks

_Tags: #Exceptions #Try-catch_

I don't know if anyone else thinks like this, but whenever I add a bit of code that I know could fail, my first instinct used to be "this needs a `try-catch`". It could be something as simple as reading from a configuration file, where the key could be missing and attempting to read it might cause an unhandled exception. The problem arises when I need to decide what exactly to do in the `catch` block. So I sometimes end up with something like this until I make up my mind (or until my code evolves enough so the way to handle this error becomes obvious):

```csharp
try
{
    var email = ConfigurationManager.AppSettings["Email"];
}
catch
{
    throw;
}
```

The thing with this is that I need to come back to it later, after my code is "done" so I can decide if this catch block serves any purpose. But sometimes, even after I've finished writing my code (or at least I finished the relevant task involving it) I still can't think of anything useful to do in case of an exception. So this is where I need to apply the rule: "don't catch an exception unless you can handle it meaningfully."

We probably can agree that rethrowing without doing anything else adds no value, so the whole try-catch block is redundant. So a better option here would be removing the whole try-catch and allowing the exception to propagate naturally, leaving the decision to the calling code. And there's also a good reason for this, other than avoiding the extra code that adds nothing: this might obscure the source of the exception, making it harder to find.

At first I was reluctant to leave error-prone code outside of a try-catch block, until I realised that the exception will be thrown anyway, so I was actually adding redundant code. "And what if someone else needs to use my code?" I would have probably asked in the past... Well, if someone needs to use my code and wants do something specific in case of an exception, then THEY should be the ones including their calling code in a try-catch block and handling the exception based on the context in which the call is made.

Even when adding something to the catch block, it might not always be a good idea. For instance, if I want to add a default value in case the key is not found in the config file, I could avoid throwing an exception by just doing something like this:

```csharp
var email = ConfigurationManager.AppSettings["Email"] ?? "default@example.com";
```

Or let's say I want to log the error and then rethrow:

```csharp
try
{
    var email = ConfigurationManager.AppSettings["Email"];
}
catch (Exception ex)
{
    Log.Error(ex.Message);
    throw;
}
```

A couple issues with this: the first one would be that I'm covering all exceptions, so whatever goes wrong it will be handled the same way. The message is not adding much context. So adding a whole try-catch block just to log the exception message without adding anything to it provides little value. But even when logging a message giving context about the error, we need to think twice as this might cause duplicate logs, since the exception is propagated (via `throw`) and a higher-level caller might be logging it again:

```csharp
// Low-level method, let's say it's called GetEmail():
try
{
    var email = ConfigurationManager.AppSettings["Email"];
}
catch (Exception ex)
{
    Log.Error("Error trying to get key 'Email' from configuration file.", ex);
    throw;
}

// Higher-level code:
try
{
    GetEmail();
}
catch (Exception ex)
{
    Log.Error("Error when getting Email.", ex);
}
```

Another no-go would be to `throw ex` instead of just `throw`, as they reset the stack trace to the current line, making it harder to identify the original source of the exception:

```csharp
try
{
    var email = ConfigurationManager.AppSettings["Email"];
}
catch (Exception ex)
{
    Log.Error("Error when trying to get key 'Email' from configuration file.", ex.Message);
    throw ex;  // Resets stack trace
}
```

All in all, these catch blocks are not even handling specific exceptions. So another rule of thumb would be to catch the most specific exception types relevant to the operation. If there's not one, you might create your own custom exception:

```csharp
public class MissingConfigurationException : Exception
{
    public MissingConfigurationException(string key)
        : base($"Configuration key '{key}' is missing.")
    { }
}
```

So, if I had to list a few guidelines that I try to follow when handling exceptions, they would be:

- Only catch exceptions when I can handle them (avoid rethrowing without adding value). Otherwise, let them propagate.
- Don't handle trivial exceptions or recoverable errors to avoid unnecessary noise. Use defaults or fallback logic instead.
- Log details that provide information about the exception. If logging is all a catch block does, consider removing it as it might be logged somewhere else.
- Use throw (not throw ex) for rethrowing
- Catch specific exception types (avoid `catch {}` or catching "Exception").
